src/components/structures/RoomView.tsx

```diff
@@ -894,14 +892,17 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
             case Action.ComposerInsert: {
                 if (payload.composerType) break;
 
-                if (this.state.searching && payload.timelineRenderingType === TimelineRenderingType.Room) {
+                if (this.state.timelineRenderingType === TimelineRenderingType.Search &&
+                    payload.timelineRenderingType === TimelineRenderingType.Search
+                ) {
                     // we don't have the composer rendered in this state, so bring it back first
                     await this.onCancelSearchClick();
                 }
 
                 // re-dispatch to the correct composer
-                dis.dispatch({
-                    ...payload,
+                dis.dispatch<ComposerInsertPayload>({
+                    ...(payload as ComposerInsertPayload),
```

Comment:
If you use a union type on the signature `payload: ActionPayload | ComposerInsertPayload | etc` then the switch will do the type assertion and the cast of payload here won't be needed.